composer/algorithms/seq_length_warmup/seq_length_warmup.py

```diff
@@ -64,25 +67,32 @@ def set_batch_sequence_length(batch: Dict[str, torch.Tensor], curr_seq_len: int,
     """
 
     assert isinstance(batch, Mapping)
+    if 'attention_mask' not in batch:
+        raise ValueError('Sequence Length Warmup requires that the batch has "attention_mask".')
 
-    if truncate:
-        for k in batch.keys():
-            batch[k] = batch[k][:, :curr_seq_len]
-    else:
-        # ensure new tensor shape is divisible by curr_seq_len
-        input_ids = batch['input_ids'].view(-1)
-        tensor_len = (input_ids.shape[0] // curr_seq_len) * curr_seq_len
+    if curr_seq_len >= batch['attention_mask'].shape[1]:
+        return batch
 
-        input_ids = input_ids[:tensor_len]
-        input_ids = input_ids.view(-1, curr_seq_len)
-        batch['input_ids'] = input_ids
+    # Truncate, but preserve end-of-sequence tokens
+    if preserve_end_of_sequence:
+        r_idx = torch.arange(batch['attention_mask'].shape[0])
+        # eos_idx should point to the final token index for each batch sample
+        eos_idx = batch['attention_mask'].sum(1).long() - 1
```

Comment:
```suggestion
        eos_idx = batch['attention_mask'].sum(dim=1).long() - 1
```
nit: use `dim` kwargs for explicit